tests/scripts/generate_test_code.py

```diff
@@ -0,0 +1,837 @@
+    noT, dep = ('!', dep[1:]) if dep[0] == '!' else ('', dep)
+    if len(dep) == 0:
+        raise GeneratorInputError("Dependency should not be an empty string.")
+    dep_check = '''
+        case {id}:
+            {{
+#if {noT}defined({macro})
+                ret = DEPENDENCY_SUPPORTED;
+#else
+                ret = DEPENDENCY_NOT_SUPPORTED;
+#endif
+            }}
+            break;'''.format(noT=noT, macro=dep, id=dep_id)
+    return dep_check
+
+
+def gen_expression_check(exp_id, exp):
+    """
+    Generates code for evaluating an integer expression using
+    associated expression Id.
+
+    :param exp_id: Expression Identifier
+    :param exp: Expression/Macro
+    :return: Expression check code
+    """
+    if exp_id < 0:
+        raise GeneratorInputError("Expression Id should be a positive "
+                                  "integer.")
+    if len(exp) == 0:
+        raise GeneratorInputError("Expression should not be an empty string.")
+    exp_code = '''
+        case {exp_id}:
+            {{
+                *out_value = {expression};
+            }}
+            break;'''.format(exp_id=exp_id, expression=exp)
+    return exp_code
+
+
+def write_deps(out_data_f, test_deps, unique_deps):
+    """
+    Write dependencies to intermediate test data file, replacing
+    the string form with identifiers. Also, generates dependency
+    check code.
+
+    :param out_data_f: Output intermediate data file
+    :param test_deps: Dependencies
+    :param unique_deps: Mutable list to track unique dependencies
+           that are global to this re-entrant function.
+    :return: returns dependency check code.
+    """
+    dep_check_code = ''
+    if len(test_deps):
+        out_data_f.write('depends_on')
+        for dep in test_deps:
+            if dep not in unique_deps:
+                unique_deps.append(dep)
+                dep_id = unique_deps.index(dep)
+                dep_check_code += gen_dep_check(dep_id, dep)
+            else:
+                dep_id = unique_deps.index(dep)
+            out_data_f.write(':' + str(dep_id))
+        out_data_f.write('\n')
+    return dep_check_code
+
+
+def write_parameters(out_data_f, test_args, func_args, unique_expressions):
+    """
+    Writes test parameters to the intermediate data file, replacing
+    the string form with identifiers. Also, generates expression
+    check code.
+
+    :param out_data_f: Output intermediate data file
+    :param test_args: Test parameters
+    :param func_args: Function arguments
+    :param unique_expressions: Mutable list to track unique
+           expressions that are global to this re-entrant function.
+    :return: Returns expression check code.
+    """
+    expression_code = ''
+    for i in range(len(test_args)):
+        typ = func_args[i]
+        val = test_args[i]
+
+        # check if val is a non literal int val (i.e. an expression)
+        if typ == 'int' and not re.match('(\d+$)|((0x)?[0-9a-fA-F]+$)', val):
```

Comment:
I am not sure the 0x should be optional. Also, I think you can move the $ out of the brackets.

Reply:
Indeed, `ABC` should not be parsed as an integer literal. And please allow `0X` as well. You can simplify this regexp a bit by adding the `re.I` flag.

Reply:
Andres's point is that you could write the regex as `r'(\d+|0x[0-9a-f]+)$'`, with the `$` at the end (this version assumes the `re.I` flag).

By the way, please use `r'…'` systematically for regular expressions. Otherwise you'll get bitten by backslash handling sooner or later.

Reply:
I understand Andres's point. But the regex does not work if move `$` out.
```
>>> re.match('(\d+$)|(0x[0-9a-fA-F]+$)', '0x456').groups()
(None, '0x456')
>>> re.match('(\d+)|(0x[0-9a-fA-F]+)$', '0x456').groups()    <--- different result
('0', None)
>>> re.match('(\d+)|(0x[0-9a-fA-F]+)$', '0xgrf').groups()     <--- Still matches!
('0', None)
>>> re.match('(\d+$)|(0x[0-9a-fA-F]+$)', '0xgrf').groups()    <--- Fails with $ inside
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'NoneType' object has no attribute 'groups'
>>>
```

Reply:
`r'(\d+|0x[0-9a-f]+)$'`, not `r'(\d+)|(0x[0-9a-f]+)$'`. In the version you tested, because of the extra parentheses, the `$` does not apply to the digits-only case.